src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.ValidateFormatString
+{
+    internal abstract class AbstractValidateFormatStringDiagnosticAnalyzer : DiagnosticAnalyzer, IBuiltInAnalyzer
+    {
+        public const string DiagnosticId = IDEDiagnosticIds.ValidateFormatStringDiagnosticID;
+        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(FeaturesResources.Invalid_format_string), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(FeaturesResources.Format_string_0_contains_invalid_placeholder_1), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(FeaturesResources.Invalid_format_string), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+
+        protected static DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticId,
+            Title,
+            MessageFormat,
+            DiagnosticCategory.Compiler,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public bool OpenFileOnly(Workspace workspace) => false;
+
+        protected static bool TryGetFormatMethod(SyntaxNodeAnalysisContext context, int numberOfArguments, SymbolInfo symbolInfo, out IMethodSymbol method)
+        {
+            method = null;
+
+            if (symbolInfo.Symbol == null)
+            {
+                return false;
+            }
+
+            if (!symbolInfo.Symbol.ContainingSymbol.Equals(context.SemanticModel.Compilation.GetTypeByMetadataName(typeof(string).FullName)))
```

Comment:
I wouldn't try to optimize for the case where this is no string type, but in general, you can split your analyzer into a `CompilationStarted` analyzer, and lookup necessary types like this, and then avoid registering any other analyzer if the type isn't present, rather than looking it up (and failing) for every invocation expression in the project.